common/transport/amqp/src/sender_link.ts

```diff
@@ -0,0 +1,229 @@
+            debug('creating sender with amqp10 for: ' + this._linkAddress);
+            this._amqp10Client.createSender(this._linkAddress, this._linkOptions)
+              .then((amqp10link) => {
+                debug('sender object created by amqp10 for endpoint: ' + this._linkAddress);
+                if (this._fsm.state === 'attaching') {
+                  this._linkObject = amqp10link;
+                  /*Codes_SRS_NODE_AMQP_SENDER_LINK_16_006: [The `SenderLink` object should subscribe to the `detached` event of the newly created `amqp10` link object.]*/
+                  this._linkObject.on('detached', this._detachHandler);
+                  /*Codes_SRS_NODE_AMQP_SENDER_LINK_16_007: [The `SenderLink` object should subscribe to the `errorReceived` event of the newly created `amqp10` link object.]*/
+                  this._linkObject.on('errorReceived', this._errorHandler);
+                  this._fsm.transition('attached', callback);
+                } else {
+                  debug('client forceDetached us already - cleaning up');
+                  amqp10link.forceDetach();
+                }
+                return null;
+              })
+              .catch((err) => {
+                debug('amqp10 failed to create sender: ' + err.toString());
+                this._fsm.transition('detached', callback, err);
+              });
+          },
+          detach: (callback) => this._fsm.transition('detaching', callback),
+          forceDetach: (err) => this._fsm.transition('detached', undefined, err),
+          send: (message, callback) => pushToQueue(message, callback)
+        },
+        attached: {
+          _onEnter: (callback) => {
+            let toSend = this._unsentMessageQueue.shift();
+            while (toSend) {
+              this._fsm.handle('send', toSend.message, toSend.callback);
+              toSend = this._unsentMessageQueue.shift();
+            }
+            if (callback) callback();
+          },
+          attach: (callback) => callback(),
+          detach: (callback) => this._fsm.transition('detaching', callback),
+          forceDetach: (err) => {
+            if (this._linkObject) {
+              this._removeListeners();
+              this._linkObject.forceDetach();
+            }
+
+            this._fsm.transition('detached', undefined, err);
+          },
+          send: (message, callback) => {
+            const op = {
+              message: message,
+              callback: callback
+            };
+
+            this._pendingMessageQueue.push(op);
+
+            /*Codes_SRS_NODE_COMMON_AMQP_16_011: [All methods should treat the `done` callback argument as optional and not throw if it is not passed as argument.]*/
+            let _processPendingMessageCallback = (error?, result?) => {
+              const opIndex = this._pendingMessageQueue.indexOf(op);
+              if (opIndex >= 0) {
+                this._pendingMessageQueue.splice(opIndex, 1);
+                if (op.callback) {
```

Comment:
>if (op.callback) { [](start = 16, length = 18)

Suggest adding a comment that we only call the callback when the message was actually found in the queue because it could have been removed by an error handler and we don't want to call the callback twice (this is important but not readily apparent).

Reply:
if the message isn't found in the queue how would we call the callback? i'm not sure I understand the comment.

---
In reply to: [139546179](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139546179) [](ancestors = 139546179)

Reply:
I think I was slightly confused and this comment resolves to "I was confused reading this.  Maybe someone else will be too."

---
In reply to: [139562522](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139562522) [](ancestors = 139562522,139546179)